Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm

```diff
@@ -1437,138 +1433,126 @@ static String mutationObserverNotificationString()
-
-    return roundedIntPoint(transformedPoint);
-}
-
-IntRect UnifiedPDFPlugin::convertFromDocumentToPlugin(const IntRect& documentSpaceRect) const
-{
-    auto transformedRect = FloatRect { documentSpaceRect };
-
-    transformedRect.scale(m_documentLayout.scale());
-    auto padding = centeringOffset();
-    transformedRect.move(padding.width(), padding.height());
-    transformedRect.scale(m_scaleFactor);
-    transformedRect.moveBy(-FloatPoint { m_scrollOffset });
-
-    return roundedIntRect(transformedRect);
-}
-
-std::optional<PDFDocumentLayout::PageIndex> UnifiedPDFPlugin::pageIndexForDocumentPoint(const WebCore::IntPoint& point) const
+std::optional<PDFDocumentLayout::PageIndex> UnifiedPDFPlugin::pageIndexForDocumentPoint(const FloatPoint& point) const
 {
     for (PDFDocumentLayout::PageIndex index = 0; index < m_documentLayout.pageCount(); ++index) {
         auto pageBounds = m_documentLayout.layoutBoundsForPageAtIndex(index);
         if (pageBounds.contains(point))
             return index;
     }
 
-    return std::nullopt;
+    return { };
 }
 
 PDFDocumentLayout::PageIndex UnifiedPDFPlugin::indexForCurrentPageInView() const
 {
-    auto centerInDocumentSpace = convertFromPluginToDocument(flooredIntPoint(size() / 2));
+    auto centerInDocumentSpace = convertDown(CoordinateSpace::Plugin, CoordinateSpace::PDFDocumentLayout, FloatPoint { flooredIntPoint(size() / 2) });
     return m_documentLayout.nearestPageIndexForDocumentPoint(centerInDocumentSpace);
 }
 
-RetainPtr<PDFAnnotation> UnifiedPDFPlugin::annotationForRootViewPoint(const WebCore::IntPoint& point) const
+RetainPtr<PDFAnnotation> UnifiedPDFPlugin::annotationForRootViewPoint(const IntPoint& point) const
 {
-    auto pointInDocumentSpace = convertFromPluginToDocument(convertFromRootViewToPlugin(point));
+    auto pointInDocumentSpace = convertDown(CoordinateSpace::Plugin, CoordinateSpace::PDFDocumentLayout, FloatPoint { convertFromRootViewToPlugin(point) });
     auto pageIndex = pageIndexForDocumentPoint(pointInDocumentSpace);
     if (!pageIndex)
         return nullptr;
 
     auto page = m_documentLayout.pageAtIndex(pageIndex.value());
-    return [page annotationAtPoint:convertFromDocumentToPage(pointInDocumentSpace, pageIndex.value())];
+    return [page annotationAtPoint:convertDown(CoordinateSpace::PDFDocumentLayout, CoordinateSpace::PDFPage, pointInDocumentSpace, pageIndex.value())];
 }
 
-IntRect UnifiedPDFPlugin::convertFromPageToRootView(const IntRect& pageRect, PDFDocumentLayout::PageIndex pageIndex) const
+template <typename T>
+T UnifiedPDFPlugin::convertUp(CoordinateSpace sourceSpace, CoordinateSpace destinationSpace, T sourceValue, std::optional<PDFDocumentLayout::PageIndex> pageIndex) const
 {
-    return convertFromPluginToRootView(convertFromDocumentToPlugin(convertFromPageToDocument(pageRect, pageIndex)));
-}
+    static_assert(std::is_same<T, FloatPoint>::value || std::is_same<T, FloatRect>::value, "Coordinate conversion should use float types");
```

Comment:
I think we can also add another assertion here, to ensure the conversion direction makes sense?

```cpp
static_assert(enumToUnderlyingType(sourceSpace) < enumToUnderlyingType(destinationSpace), <some sensible message about moving up>);
static_assert(std::is_same<T, FloatPoint>::value || std::is_same<T, FloatRect>::value, "Coordinate conversion should use float types");
```

Reply:
I tried this but got `static assertion expression is not an integral constant expression`.